pkg/operator/quorumguardcontroller/quorumguardcontroller.go

```diff
@@ -0,0 +1,142 @@
+package quorumguardcontroller
+
+import (
+	"context"
+	"time"
+
+	configclientv1 "github.com/openshift/client-go/config/clientset/versioned/typed/config/v1"
+	"github.com/openshift/cluster-etcd-operator/pkg/operator/etcd_assets"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceread"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	configv1 "github.com/openshift/api/config/v1"
+	operatorv1 "github.com/openshift/api/operator/v1"
+	configv1informers "github.com/openshift/client-go/config/informers/externalversions/config/v1"
+	configv1listers "github.com/openshift/client-go/config/listers/config/v1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/klog/v2"
+
+	"github.com/openshift/cluster-etcd-operator/pkg/operator/operatorclient"
+)
+
+const (
+	EtcdGuardDeploymentName   = "etcd-quorum-guard"
+	infrastructureClusterName = "cluster"
+)
+
+// watches the etcd quorum guard deployment, create if not exists
+type QuorumGuardController struct {
+	operatorClient v1helpers.OperatorClient
+	kubeClient     kubernetes.Interface
+	podLister      corev1listers.PodLister
+	nodeLister     corev1listers.NodeLister
+	networkLister  configv1listers.NetworkLister
+	infraClient    configclientv1.InfrastructuresGetter
+	haMode         configv1.HighAvailabilityMode
+}
+
+func NewQuorumGuardController(
+	operatorClient v1helpers.OperatorClient,
+	kubeClient kubernetes.Interface,
+	kubeInformers v1helpers.KubeInformersForNamespaces,
+	networkInformer configv1informers.NetworkInformer,
+	eventRecorder events.Recorder,
+	infraClient configclientv1.InfrastructuresGetter,
+) factory.Controller {
+	c := &QuorumGuardController{
+		operatorClient: operatorClient,
+		kubeClient:     kubeClient,
+		podLister:      kubeInformers.InformersFor(operatorclient.TargetNamespace).Core().V1().Pods().Lister(),
+		nodeLister:     kubeInformers.InformersFor("").Core().V1().Nodes().Lister(),
+		networkLister:  networkInformer.Lister(),
+		infraClient:    infraClient,
+	}
+	return factory.New().ResyncEvery(time.Minute).WithInformers(
+		kubeInformers.InformersFor(operatorclient.TargetNamespace).Core().V1().Pods().Informer(),
+		kubeInformers.InformersFor("").Core().V1().Nodes().Informer(),
+		networkInformer.Informer(),
+		operatorClient.Informer(),
+	).WithSync(c.sync).ResyncEvery(time.Minute).ToController("QuorumGuardController", eventRecorder.WithComponentSuffix("quorum-guard-controller"))
+}
+
+func (c *QuorumGuardController) isInHAMode(ctx context.Context) (bool, error) {
+	// update once
+	if c.haMode == "" {
+		infraData, err := c.infraClient.Infrastructures().Get(ctx, infrastructureClusterName, metav1.GetOptions{})
```

Comment:
I outlined my concerns[1] with using this API.

[1] https://github.com/openshift/enhancements/pull/555#pullrequestreview-562743270

Reply:
do we have an alternative if this API does not merge?

Reply:
not sure. We can check number of masters ?

Reply:
we can not deploy etcd guard till we have only one master

Reply:
then we don't need to validate this flag at all